### PR TITLE
feat: add ak.from_cudf and rewrite to_cudf via libcudf

### LINKED ISSUE #3605: Use of cudf Column internals
I'm opening this issue to follow up on https://github.com/scikit-hep/awkward/pull/3328 and https://github.com/rapidsai/cudf/issues/17483. I know that awkward array has relied on some of cuDF's internal Column APIs in the past. cuDF Is now attempting to provide better low-level APIs via the pylibcudf library to allow direct interaction with our C++ primitives. We'll eventually be working on exposing a more easily pluggable middle layer (i.e. to be able to deal with objects in our pandas-like layer), but for now I'm hoping that the pylibcudf bindings are sufficient for a lot of what awkward needs. It is sufficient to call `to_pylibcudf` on cudf.Series/DataFrame objects to get the underlying pylibcudf types, and the reverse can be done with `from_pylibcudf`.

Could you please let us know if those types are providing the accessors and methods that you need to support your use cases? That includes however you were using `Series._from_column` in https://github.com/rapidsai/cudf/issues/17483, but also anything else that you might currently be reaching into our internals for. We'd love to provide suitable public APIs to help you all out now that we're far enough along in our internals refactor to support that.

### PR DESCRIPTION
**Pull request by aashirvad08**
## Summary

Adds `ak.from_cudf(series)` — a new public function that converts a
`cudf.Series` into an `ak.Array` on GPU using only public pylibcudf APIs,
with no device-to-host copy and no cuDF internal APIs.

## Motivation

The existing `ak.to_cudf` path depends on `Series._from_column`, a private
cuDF API tied to column internals that are evolving as part of the cuDF
column refactor [(rapidsai/cudf#18726).](https://github.com/rapidsai/cudf/issues/18726) 
A stable public alternative is now available via `pylibcudf`, together with the Arrow C Device Interface.

This PR implements the Awkward side of that migration.

Conversion path:
series.to_pylibcudf()
  └─ recurse(col) dispatches on col.type().id()
       ├─ primitive → NumpyArray via col.data_buffer() + cp.asarray()
       ├─ bool8     → NumpyArray (byte → bool cast)
       ├─ list      → ListOffsetArray via col.offsets() + col.child(0)
       ├─ struct    → RecordArray via col.child(i) for each field
       └─ string    → ListOffsetArray[char] via col.child(0) + col.data_buffer()
  └─ _finalize() wraps nullable columns with BitMaskedArray.simplified()

Related to #3328 
Closes #3605

### CODE REVIEW

**Comment by martindurant** [comment]
Thanks for contributing!

> converts a flat numeric cudf.Series into an ak.Array

I'm not sure how useful this is in the awkward context, where the whole point is to have nested/var types. I would say that the natural representation of a cudf Series for numpy-like operations is cupy. It is hard to follow the nested recursive schema with the new column API? Can the existing arrow conversion code be applied?

Ref: https://github.com/intake/akimbo/blob/main/src/akimbo/ak_from_cudf.py

> Import nanoarrow (required to consume the PyCapsule protocol)

Other transformers already depend on pyarrow, can we not use that?

**Comment by aashirvad08** [comment]
hey,
>converts a flat numeric cudf.Series into an ak.Array

agreed , I kept the scope here minimal to first sort migration from Series._from_column
I am happy to extend this to support nested types 

>Can the existing arrow conversion code be applied?

ak.from_arrow already handles full Arrow schemas  but using series.to_arrow() would introduce a device-to-host copy, losing zero-copy GPU path
If zero-copy was  not required then switching to to_arrow() + ak.from_arrow would have simplify the implementation and support all types 

>Import nanoarrow (required to consume the PyCapsule protocol)

nanoarrow was chosen cause again  it can consume the __arrow_c_device_array__ capsule  without copying although pyarrow operates on host buffers
if a D2H transfer is acceptable we can drop nanoarrow and rely on pyarrow + ak.from_arrow

**Comment by martindurant** [comment]
> would introduce a device-to-host copy

I didn't mean literally using that code, but the pattern in that code, recursing into the schema structure. That's what my code in `akimbo` does, by digging into the "columns" structure. It was "inspired by" the from_arrow code.

At the time, there were sufficient differences between the "columns" structures and CPU arrow, that the from_arrow code could not be reused. If those differences have reduced, then the code could be reused (by copy&modify or somehow abstracting out the buffers). If not, then nanoarrow won't help for any complex case, and so is not really useful. 

> it can consume the arrow_c_device_array capsule without copying

Essentially cupy for flat types, no? We already depend on that for GPUs.

**Comment by aashirvad08** [comment]
Good point.
I found `studies/cudf-to-awkward.py` in the repo which is exactly the pattern you're describing--a recurse() function mirroring from_arrow but operating on cuDF column internals via cupy.asarray(), covering structs, lists, strings, etc so  no nanoarrow needed

My hesitation is that it relies on column.base_data, base_children, and base_mask, which look internal and may shift as part of [rapidsai/cudf#18726.](https://github.com/rapidsai/cudf/issues/18726) 
My original implementation tried to stay within the public pylibcudf / Arrow interface for that reason but I agree the recursive approach is the right direction and you're correct that for flat types nanoarrow is just cp.asarray()  which cupy  already provides

I'll rework the  PR around that study pattern, dropping nanoarrow and adding full nested/list/struct/null support.
Before doing that, do you know if `base_data`, `base_children`, and `base_mask`
are still the appropriate accessors in cuDF ≥ 25.02, or if they’ve changed as
part of the column refactor?

**Comment by martindurant** [comment]
> My hesitation is that it relies on column.base_data, base_children, and base_mask, which look internal and may shift as part of https://github.com/rapidsai/cudf/issues/18726

Yes, that is exactly the situation. With the internal restructuring in cudf, these accessor attributes and methods will find permanent places to live, so we should be able to implement the conversion in the same manner as before but for the new layout. I don't think *too* much has changed, except that from the perspective of `akimbo`, which also cares about string and time operations, more work is needed.

**Comment by aashirvad08** [comment]
Great, thanks for confirming. I'll update this PR to follow the `studies/cudf-to-awkward.py` recursive pattern, adapt it to the new pylibcudf layout, drop nanoarrow, and cover the full type set from the start.

File: src/awkward/_connect/cudf.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    nested types raise ``NotImplementedError``.
+    """
+
+    try:
+        import cudf
+    except ImportError as err:
+        raise ImportError(
+            """to use ak.from_cudf, you must install the 'cudf' package with:
+
+    pip install cudf-cu13
```
Comment (vyasr) [inline src/awkward/_connect/cudf.py]: Not sure if you want to be prescriptive about the CUDA version here, but I'll defer to other reviewers on what awkward prefers.

**Comment by vyasr** [comment]
Sorry for the slow response here. 

> My hesitation is that it relies on column.base_data, base_children, and base_mask, which look internal and may shift as part of https://github.com/rapidsai/cudf/issues/18726

To be explicit, those attributes are all gone now.

> I didn't mean literally using that code, but the pattern in that code, recursing into the schema structure. That's what my code in akimbo does, by digging into the "columns" structure. It was "inspired by" the from_arrow code.
> 
> At the time, there were sufficient differences between the "columns" structures and CPU arrow, that the from_arrow code could not be reused. If those differences have reduced, then the code could be reused (by copy&modify or somehow abstracting out the buffers).

Yes, this should work now. Concretely, an if-else cascade (or a value-based singledispatch-like approach) based on the value of [pylibcudf_column.type().id()](https://github.com/rapidsai/cudf/blob/bd34658311e4b7319d8296317d9bc8938c6bfb2d/python/pylibcudf/pylibcudf/types.pyx#L124) should allow you to traverse arrow structures in the same way you would expect for CPU arrow. The main edge cases you'll run into are unspecified corners of the Arrow spec itself such as whether empty list columns have one offset or none, or whether null masks can be unallocated altogether for empty columns, etc.

**Comment by aashirvad08** [comment]
Thanks for the update — good to know those attributes are gone, and helpful to have the concrete direction. I'll implement the traversal based on plc_column.type().id(), following the same if-else cascade as from_arrow but reading buffers from pylibcudf directly via CuPy. I'll handle the Arrow edge cases (empty list offsets, unallocated null masks) as I go and note anything that needs clarification from the cuDF side. Will push an updated version.

**Comment by aashirvad08** [comment]
The failing test test_2616_use_pyarrow_for_strings.py::test_split_whitespace is unrelated to this PR — none of my changes touch the string operations path. It doesn't reproduce locally with multiple pyarrow versions.

**Comment by ikrommyd** [comment]
> The failing test test_2616_use_pyarrow_for_strings.py::test_split_whitespace is unrelated to this PR — none of my changes touch the string operations path. It doesn't reproduce locally with multiple pyarrow versions.

this is a known flaky test

**Comment by aashirvad08** [comment]
any updates ?

**Comment by martindurant** [comment]
I'm sorry, I didn't realise there were new changes here!

A quick glance suggests it is going well - I see a bunch of decent looking tests. Perhaps we should have even more coverage of deep/nested types and something across the various times units and such. I remember that in arrow, there is a difference between nullable, but mask=None versus nullable but happens to have no nulls - I don't know if that's the case for cudf, but we should explicitly test it.

**Comment by aashirvad08** [comment]
Thanks I’ll add tests accordingly

**Comment by martindurant** [comment]
@ianna , since you are thinking about extending GPU capabilities, do you have anything o add to a more concerted push for cuDF integration? I am thinking that the most concrete upside is being able to read/write parquet directly with GPU memory. However, some things are more cleanly implemented in cudf (i.e., counterparts of arrow.compute) which might be handy; I am thinking string and time/date manipulations.

**Review by vyasr — COMMENTED** [review]
Just realized I never committed this comment from weeks ago, sorry...

**Comment by vyasr** [comment]
> I'm sorry, I didn't realise there were new changes here!
> 
> A quick glance suggests it is going well - I see a bunch of decent looking tests. Perhaps we should have even more coverage of deep/nested types and something across the various times units and such. I remember that in arrow, there is a difference between nullable, but mask=None versus nullable but happens to have no nulls - I don't know if that's the case for cudf, but we should explicitly test it.

There is a similar situtation in cudf as well, but it is subtly different. In cudf you can distinguish between whether [a null mask is present at all](https://github.com/rapidsai/cudf/blob/cef175643e815df1555df7fb6ef2374c8c1e54e0/python/pylibcudf/pylibcudf/column.pyx#L1343) vs whether [the null count is zero](https://github.com/rapidsai/cudf/blob/cef175643e815df1555df7fb6ef2374c8c1e54e0/python/pylibcudf/pylibcudf/column.pyx#L1355). The null count is always zero if there is no mask, but you could also have a mask with all valid elements. In our representation ["nullable" is an indication of whether a column's null mask has been allocated at all](https://github.com/rapidsai/cudf/blob/cef175643e815df1555df7fb6ef2374c8c1e54e0/cpp/include/cudf/column/column.hpp#L207).

File: src/awkward/_connect/cudf.py (hunk 1 above)
Comment (ianna) [reply src/awkward/_connect/cudf.py]: Agree. We still support both CUDA 12 and 13, not 11.

**Review by ianna — COMMENTED** [review]
@aashirvad08 - thanks for working on this! I will let @martindurant to have the final word on it 👍

File: src/awkward/operations/ak_from_cudf.py [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+    cp = _cupy()
+
+    if buf is None:
+        return cp.empty(0, dtype=dtype)
+
+    dtype = np.dtype(dtype)
+    mem = cp.cuda.UnownedMemory(ptr=int(buf.ptr), size=int(buf.size), owner=buf)
+    memptr = cp.cuda.MemoryPointer(mem, 0)
+    length = int(buf.size) // dtype.itemsize
+    return cp.ndarray(length, dtype=dtype, memptr=memptr)
```
Comment (martindurant) [inline src/awkward/operations/ak_from_cudf.py]: `memptr` is certainly allowed, I don't know why the linter is complaining. Perhaps it doesn't know the difference between np anc cp.

File: src/awkward/operations/ak_from_cudf.py [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+    Missing buffers represent empty columns in pylibcudf, so they become
+    zero-length CuPy arrays of the requested dtype.
+    """
+    cp = _cupy()
+
+    if buf is None:
+        return cp.empty(0, dtype=dtype)
+
+    dtype = np.dtype(dtype)
+    mem = cp.cuda.UnownedMemory(ptr=int(buf.ptr), size=int(buf.size), owner=buf)
```
Comment (martindurant) [inline src/awkward/operations/ak_from_cudf.py]: Does `owner=buf` ensure this buffer doesn't get cleaned up?

File: src/awkward/operations/ak_from_cudf.py [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+    """
+    Wrap a pylibcudf gpumemoryview as a CuPy array without copying.
+
+    Missing buffers represent empty columns in pylibcudf, so they become
+    zero-length CuPy arrays of the requested dtype.
+    """
+    cp = _cupy()
+
+    if buf is None:
+        return cp.empty(0, dtype=dtype)
```
Comment (martindurant) [inline src/awkward/operations/ak_from_cudf.py]: Why can't we use None as the return value?

File: codecov.yml [hunk 5]
```diff
@@ -20,3 +20,7 @@ coverage:
 
 github_checks:
   annotations: false
+
+ignore:
+  - "src/awkward/_connect/cudf.py"
+  - "src/awkward/operations/ak_from_cudf.py"
```
Comment (martindurant) [inline codecov.yml:26]: How much do we miss?

File: src/awkward/_connect/cudf.py [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+        ) from err
+
+    return result[0] if isinstance(result, tuple) else result
+
+
+def from_cudf(series):
+    """
+    Args:
+        series (cudf.Series): The cuDF Series to convert into a low-level
+            Awkward layout.
```
Comment (martindurant) [inline src/awkward/_connect/cudf.py]: We may also wish to accept dataframes, which are equivalent to a series with top-level record structure. The only difference is possible arrow metadata on the columns. `akimbo` does this somewhere.

(can be left for the future)

File: src/awkward/_connect/cudf.py [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+
+    Converts a cuDF Series into a low-level Awkward layout by recursively
+    traversing the corresponding pylibcudf column.
+
+    Primitive, boolean, list, struct, string, and nullable columns are
+    supported. Other column types raise ``NotImplementedError``.
+    """
+    try:
+        import cudf
```
Comment (martindurant) [inline src/awkward/_connect/cudf.py:31]: We need cudf, libcudf and cupy here, but we still have import libcudf and _cudf() in the other module. Probably we should do all of them just once as a helper function. It's probably OK if we can't import this code at all without them.

File: src/awkward/_connect/cudf.py [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+        target_dtype = cp.dtype(dtype)
+        if data.dtype == target_dtype:
+            return data
+
+        if data.dtype.itemsize == target_dtype.itemsize:
+            return data.view(target_dtype)
+
+        return cp.asarray(buffer, dtype=target_dtype)
+
+    def _data_to_cupy(col, dtype):
```
Comment (martindurant) [inline src/awkward/_connect/cudf.py]: Type annotations would be handy on these helpers, it will avoid errors in the future

File: src/awkward/_connect/cudf.py [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+            return Index64(offsets)
+
+        if offsets.dtype == cp.dtype("int32"):
+            return Index32(offsets)
+        elif offsets.dtype == cp.dtype("int64"):
+            return Index64(offsets)
+        else:
+            return Index64(cp.asarray(offsets, dtype="int64"))
+
+    def _struct_fields(col, dtype, num_children):
```
Comment (martindurant) [inline src/awkward/_connect/cudf.py]: This is really annoying! Is there no helper in cudf? I don't really understand the possibilities.

File: src/awkward/_connect/cudf.py [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+        return names, child_dtypes[:num_children]
+
+    def _finalize(layout, col):
+        offset = _column_offset(col)
+        size = _column_size(col)
+        stop = offset + size
+
+        if layout.length < stop:
+            raise RuntimeError(
+                "pylibcudf column buffers have an unexpected shape for ak.from_cudf"
```
Comment (martindurant) [inline src/awkward/_connect/cudf.py]: Is there a case like this?

File: src/awkward/_connect/cudf.py [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+                    length=stop,
+                    lsb_order=True,
+                )
+
+        if offset != 0 or layout.length != size:
+            layout = layout[offset:stop]
+
+        return layout
+
+    def recurse(col, dtype=None):
```
Comment (martindurant) [inline src/awkward/_connect/cudf.py]: I don't love having this (and other functions) be dynamically recreated in the outer function. It's OK to have NameErrors if imports fail.

File: src/awkward/operations/__init__.py [hunk 12]
```diff
@@ -36,6 +36,7 @@
 from awkward.operations.ak_from_avro_file import *
 from awkward.operations.ak_from_buffers import *
 from awkward.operations.ak_from_categorical import *
+from awkward.operations.ak_from_cudf import from_cudf
```
Comment (martindurant) [inline src/awkward/operations/__init__.py]: You did deffine `__all__`, so `*` would be fine here to match the other imports.

File: src/awkward/operations/ak_from_cudf.py [hunk 13]
```diff
@@ -0,0 +1,359 @@
...
+
+
+def _column_offset(col):
+    offset = getattr(col, "offset", 0)
+    return offset() if callable(offset) else offset
+
+
+def _column_null_count(col):
+    null_count = col.null_count
+    return null_count() if callable(null_count) else null_count
```
Comment (martindurant) [inline src/awkward/operations/ak_from_cudf.py]: This pattern is very common

File: src/awkward/operations/ak_from_cudf.py [hunk 14]
```diff
@@ -0,0 +1,359 @@
...
+
+
+def _num_children(col):
+    num_children = getattr(col, "num_children", None)
+    if callable(num_children):
+        return num_children()
+    elif num_children is not None:
+        return num_children
+    else:
+        return len(col.children())
```
Comment (martindurant) [inline src/awkward/operations/ak_from_cudf.py]: This also seems to crop up a lot

File: src/awkward/operations/ak_from_cudf.py [hunk 15]
```diff
@@ -0,0 +1,359 @@
...
+    return names[:num_children]
+
+
+def _to_pylibcudf_column(series):
+    try:
+        result = series.to_pylibcudf()
+    except AttributeError as err:
+        raise RuntimeError(
+            "cudf.Series.to_pylibcudf() is required by ak.from_cudf. "
+            "Please use cudf >= 25.02."
```
Comment (martindurant) [inline src/awkward/operations/ak_from_cudf.py]: Don't we already check on import?

File: src/awkward/operations/ak_from_cudf.py [hunk 16]
```diff
@@ -0,0 +1,359 @@
...
+
+    return _finalize(layout, col)
+
+
+def _dataframe_to_layout(dataframe):
+    fields = list(dataframe.columns)
+    contents = [
+        _column_to_layout(_to_pylibcudf_column(dataframe[name])) for name in fields
+    ]
+    return RecordArray(contents, fields, length=len(dataframe))
```
Comment (martindurant) [inline src/awkward/operations/ak_from_cudf.py]: OK, sorry, you do handle dataframes

File: src/awkward/operations/ak_from_cudf.py [hunk 17]
```diff
@@ -0,0 +1,359 @@
...
+
+    out = {}
+    for name, dtype in names_and_dtypes:
+        type_id = _type_id(plc, name)
+        if type_id is not None:
+            out[type_id] = dtype
+    return out
+
+
+def _column_to_layout(col):
```
Comment (martindurant) [inline src/awkward/operations/ak_from_cudf.py]: What's the difference between work done here and in _conect.cudf.recurse ?

**Review by martindurant — COMMENTED** [review]
The shape of everything here looks good. I had some confusion about the split of functionality between _connect.cudf and ak_from_cudf modules , and several smaller questions.

File: src/awkward/operations/ak_from_cudf.py (hunk 3 above)
Comment (aashirvad08) [reply src/awkward/operations/ak_from_cudf.py]: Yes, owner=buf holds a reference to the pylibcudf gpumemoryview for the CuPy allocation lifetime , ill add a comment for this explicitly

File: src/awkward/operations/ak_from_cudf.py (hunk 4 above)
Comment (aashirvad08) [reply src/awkward/operations/ak_from_cudf.py]: zero-length cupy array keeps layout uniform and returning none  would require each caller to handle empty buffers separately

File: src/awkward/_connect/cudf.py (hunk 6 above)
Comment (aashirvad08) [reply src/awkward/_connect/cudf.py]: Already did   cudf.DataFrame builds a top-level RecordArray preserving column order

File: src/awkward/_connect/cudf.py (hunk 7 above)
Comment (aashirvad08) [reply src/awkward/_connect/cudf.py:31]: I'll add a single _ensure_deps() helper in _connect/cudf.py that handles all  imports

File: src/awkward/_connect/cudf.py (hunk 8 above)
Comment (aashirvad08) [reply src/awkward/_connect/cudf.py]: sure , I'll add some

File: src/awkward/_connect/cudf.py (hunk 9 above)
Comment (aashirvad08) [reply src/awkward/_connect/cudf.py]: The possibilities seem to be `col.child_names()` on newer pylibcudf, or falling back to schema/dtype-derived names when that metadata isn’t exposed.

File: src/awkward/_connect/cudf.py (hunk 10 above)
Comment (aashirvad08) [reply src/awkward/_connect/cudf.py]: this would not happen on a well formed pylibcudf column , its a check for sliced columns or unexpected buffer or offsets

File: src/awkward/_connect/cudf.py (hunk 11 above)
Comment (aashirvad08) [reply src/awkward/_connect/cudf.py]: ok I'll then move all recursive helpers to module scope in _connect/cudf.py

File: src/awkward/operations/__init__.py (hunk 12 above)
Comment (aashirvad08) [reply src/awkward/operations/__init__.py]: yep I'll switch to from awkward.operations.ak_from_cudf import *

File: src/awkward/operations/ak_from_cudf.py (hunk 13 above)
Comment (aashirvad08) [reply src/awkward/operations/ak_from_cudf.py]: I'll extract a  helper function for this

File: src/awkward/operations/ak_from_cudf.py (hunk 15 above)
Comment (aashirvad08) [reply src/awkward/operations/ak_from_cudf.py]: this check is specifically for Series.to_pylibcudf() being version-gated

File: src/awkward/operations/ak_from_cudf.py (hunk 17 above)
Comment (aashirvad08) [reply src/awkward/operations/ak_from_cudf.py]: you're right ,I’ll consolidate this into `_connect/cudf.py` so there is only one path

File: src/awkward/operations/ak_from_cudf.py (hunk 2 above)
Comment (aashirvad08) [reply src/awkward/operations/ak_from_cudf.py]: memptr is valid for cp this is moslty  a false positive from conflicting it with np

Comment (martindurant) [reply src/awkward/operations/ak_from_cudf.py]: probably because cupy isn't explicitly imported here, only via the function call?

Comment (martindurant) [reply src/awkward/operations/ak_from_cudf.py]: (since the linter is required to pass, you'll need to add an ignore label here)

**Comment by martindurant** [comment]
Please ping me when I should review again

**Comment by aashirvad08** [comment]
yep ill pin when I push changes

File: src/awkward/_connect/cudf.py (hunk 7 above)
Comment (vyasr) [reply src/awkward/_connect/cudf.py:31]: To be clear you should never need to import `libcudf`. You should only ever need to import `pylibcudf` or `cudf` depending on what you're doing. I don't see any imports of `libcudf` right now so I'm assuming that isn't actually an issue, just letting you know.

File: src/awkward/_connect/cudf.py (hunk 10 above)
Comment (vyasr) [reply src/awkward/_connect/cudf.py]: Please feel free to open bugs if you find that we produce a column you consider ill-formed.

**Comment by martindurant** [comment]
Would you mind running pytest with coverage on the new module? We don't need to cover it all, but it would be good to record where we're up to.

**Comment by aashirvad08** [comment]
sure, will do

**Comment by martindurant** [comment]
I am hoping to get some time this week on my GPU to try this out together with akimbo - thank you for waiting!

**Comment by aashirvad08** [comment]
sounds good , let me know if u need any help from my side

**Comment by martindurant** [comment]
I am finding that there have been some upstream changes in [[py]lib]cudf 26.06.01 ... I have some running patches, but this may take some work.

**Comment by martindurant** [comment]
See https://github.com/aashirvad08/awkward/pull/1 for some changes that appear necessary (not coded nicely, just to demonstrate). 

I see that the existing to_cudf is version guarded because pylibcudf was still expected when that came out. I can look at it - that side will be needed too for anything in `akimbo` to work.

**Comment by aashirvad08** [comment]
thanks for putting this together , ill go through it and push the updated version

**Comment by martindurant** [comment]
I think I'm done with my changes, and coverage should be better here now (so no need to ignore in the config). The diff in my PR looks huge because it includes upstream changes. If you update this branch, it should look far saner.

**Comment by martindurant** [comment]
@aashirvad08 , where are we up to with this? I had a change of job in the meanwhile, but I should be able to put some time in here.

**Comment by aashirvad08** [comment]
@martindurant I’ve been quite busy over the past few weeks give me 1–2 days,  I’ll get back to you with an update on the status

**Comment by aashirvad08** [comment]
@martindurant  dropping cuDF <25.12 support is intentional and  I have also fixed the parse_version issue

**Comment by martindurant** [comment]
The codecov is missing GPU runs, right?

**Comment by aashirvad08** [comment]
GPU tests 12,13 were there

**Comment by martindurant** [comment]
Recommend changing the title to: "feat: add ak.from_cudf and rewrite to_cudf via libcudf"

**Comment by martindurant** [comment]
The following diff fixes the coverage and a couple of mask holes it uncovered:
```diff
--- a/src/awkward/contents/listoffsetarray.py
+++ b/src/awkward/contents/listoffsetarray.py
@@ -1788,9 +1788,9 @@ class ListOffsetArray(ListOffsetMeta[Content], Content):
         offsets_col = as_column(index)
         # Build the packed bitmask as a CuPy array (LSB order, 64-byte aligned)
         if mask is not None:
-            m_arr = np._module.packbits(mask, bitorder="little")
+            m_arr = numpy.packbits(mask, bitorder="little")
             if m_arr.nbytes % 64:
-                m_arr = cupy.resize(m_arr, ((m_arr.nbytes // 64) + 1) * 64)
+                m_arr = cupy._module.resize(m_arr, ((m_arr.nbytes // 64) + 1) * 64)
             m_arr = cupy.asarray(m_arr)
         else:
             m_arr = None
@@ -1846,9 +1846,7 @@ class ListOffsetArray(ListOffsetMeta[Content], Content):
         list_dt = cudf.ListDtype(cont.dtype)
         list_col = ColumnBase.create(plc_col, list_dt)
         if m_arr is not None:
-            null_count = int(
-                length - int(cupy.unpackbits(m_arr, bitorder="little")[:length].sum())
-            )
+            null_count = len(self) - mask.sum()
             return list_col.set_mask(as_buffer(m_arr), null_count)
         return list_col
 
diff --git a/src/awkward/contents/numpyarray.py b/src/awkward/contents/numpyarray.py
index c70603d5..258e1dc7 100644
--- a/src/awkward/contents/numpyarray.py
+++ b/src/awkward/contents/numpyarray.py
@@ -1190,7 +1190,6 @@ class NumpyArray(NumpyMeta, Content):
         from cudf.core.column.column import ColumnBase
         from cudf.utils.dtypes import dtype_to_pylibcudf_type
         from pylibcudf.gpumemoryview import gpumemoryview
-        from rmm.pylibrmm.device_buffer import DeviceBuffer
 
         (raw_data,) = maybe_materialize(self._data)
 
@@ -1226,13 +1225,11 @@ class NumpyArray(NumpyMeta, Content):
                 plc_col = plc.unary.cast(plc_col, dtype_to_pylibcudf_type(np_dtype))
 
         if mask is not None:
-            m = cupy.packbits(cupy.asarray(mask), bitorder="little")
+            m = cupy._module.packbits(cupy.asarray(mask), bitorder="little")
             if m.nbytes % 64:
-                m = cupy.resize(m, ((m.nbytes // 64) + 1) * 64)
-            null_count = int(
-                length - int(cupy.unpackbits(m, bitorder="little")[:length].sum())
-            )
-            mask_gmv = gpumemoryview(DeviceBuffer.from_cuda_array_interface(m))
+                m = cupy._module.resize(m, ((m.nbytes // 64) + 1) * 64)
+            null_count = len(self) - mask.sum()
+            mask_gmv = gpumemoryview(m)
             plc_col = plc_col.with_mask(mask_gmv, null_count)
 
         return ColumnBase.from_pylibcudf(plc_col)
diff --git a/tests-cuda/test_3946_to_cudf_extended.py b/tests-cuda/test_3946_to_cudf_extended.py
index 4ee975a9..51c1436f 100644
--- a/tests-cuda/test_3946_to_cudf_extended.py
+++ b/tests-cuda/test_3946_to_cudf_extended.py
@@ -2,6 +2,7 @@
 
 """Extended tests for ak.to_cudf covering:
 - All primitive dtypes (int8/16/32/64, uint8/16/32/64, float32/64, bool)
+- Direct NumpyArray and ListOffsetArray conversion, including explicit masks
 - Datetime and timedelta dtypes
 - Empty arrays of various types
 - Deeply nested structs (3+ levels)
@@ -129,6 +130,42 @@ def test_to_cudf_bool():
     assert out.dtype == np.dtype("bool")
 
 
+# ---------------------------------------------------------------------------
+# Direct layout conversion with explicit masks
+# ---------------------------------------------------------------------------
+
+
+def test_numpyarray_to_cudf_with_mask():
+    layout = ak.contents.NumpyArray(np.array([10, 99, 30, 99], dtype=np.int32))
+    mask = np.array([True, False, True, False])
+
+    column = layout._to_cudf(cudf, mask, len(layout))
+    out = cudf.Series._from_column(column)
+
+    assert _to_arrow_list(out) == [10, None, 30, None]
+
+
+def test_listoffsetarray_to_cudf_with_mask():
+    layout = ak.contents.ListOffsetArray(
+        ak.index.Index64(np.array([0, 2, 3, 5], dtype=np.int64)),
+        ak.contents.NumpyArray(np.array([10, 20, 30, 40, 50], dtype=np.int32)),
+    )
+    mask = np.array([True, False, True])
+
+    column = layout._to_cudf(cudf, mask, len(layout))
+    out = cudf.Series._from_column(column)
+
+    assert _to_arrow_list(out) == [[10, 20], None, [40, 50]]
+
+    string_layout = ak.Array(["one", "two", "three"]).layout
+    column = string_layout._to_cudf(cudf, mask, len(string_layout))
+    out = cudf.Series._from_column(column)
+
+    assert out.iloc[0] == "one"
+    assert out.isnull().iloc[1]
+    assert out.iloc[2] == "three"
+
+
```

**Comment by martindurant** [comment]
There was also an issue of converting a masted-string array back to arrow (size 0 buffer) which we can probably ignore.

Otherwise, I approve this PR!

**Comment by aashirvad08** [comment]
@martindurant  anything should I update in here ?

**Comment by martindurant** [comment]
Yes, I think the diff above, or something like it, should be added to prevent regression on the small mask operations I identified. I don't know why codecov is complaining, the coverage is pretty good.

**Comment by aashirvad08** [comment]
hey @martindurant t he coverage check shows 0% because all 62 new lines live in the Content._to_cudf methods and also  cudf isn't installed in any CI job

**Comment by martindurant** [comment]
Huh, I really should have noticed that! :)  Let's add it to the cuda environment! It needs the "rapidsai" channel (I believe a migration to conda-forge is coming soon).

**Comment by martindurant** [comment]
Unfortunately, this is showing test errors :|

**Comment by martindurant** [comment]
Since this was a force-push, I'm not quite sure what changed here. We are back to missing coverage?

**Comment by aashirvad08** [comment]
Sorry, the force push was just a commit-message amend so there were no code changes

**Comment by aashirvad08** [comment]
For coverage the GPU tests pass but they don't run with --cov or upload to codecov
The coverage job can't install cudf because the micromamba solve fails so those tests get skipped and the 62 _to_cudf lines appear uncovered

**Comment by martindurant** [comment]
> The coverage job can't install cudf because the micromamba solve fails so those tests

Can we add cov back in using pip instead, then (after the main mamba solve)?

**Comment by aashirvad08** [comment]
Hey , the coverage job is currently on python version 3.10 and cudf-cu13 requires >=3.11  
and also installing cudf using pip will invite several dependencies like pandas, numpy, cupy and numba
I was thinking of  to let the main mamba solve finish first and then install only the cudf  pieces with pip(without letting pip replace the dependencies that mamba already solved )

**Comment by ianna** [comment]
hi @aaronj0 -- When running ROOT through `cppyy` on Python 3.13, constructing or implicitly converting Python strings to `ROOT::TString` causes memory corruption inside `libcppyy_backend.so`, ultimately leading to a segmentation fault in `__memmove_avx512_unaligned_erms` during a `free()` call.

This appears to be an ABI incompatibility between `cppyy/ROOT` and `Python 3.13`, specifically in the handling of `TString` arguments and conversions.

The crash occurs during `pytest` collection, but the failure is independent of pytest — it is triggered by cppyy’s constructor and argument conversion logic.

**Comment by martindurant** [comment]
(@ianna : the last comment belongs here?)


I was not suggesting to installcudf with pip. If I understood the report above  right, mamba is solving the cudf-containing environment OK (but I haven't checked versions), but it's the addition of codecov that breaks it for some reason. That latter thing should be easy to add on with pip, though, so it might be the easiest path.

**Review by ianna — APPROVED** [review]
@aashirvad08 and @martindurant -- Thank you both -- all is green! I'll merge it now.
